### DISCUSSION TITLE
Generated class named "Object" is being confused with JavaScript Object

### DISCUSSION THREAD

**Discussion by Goncalo-Marques** [wailsapp/wails#1982] [Q&A]
## Context

I am having some problems with my generated models and want to understand if it represents a bug or if there is a workaround.

## Problem

Basically, I have a Go method that should return a slice of a structure named `Object`:
```go
type Object struct {}

type App struct {
	ctx context.Context
}

func (a *App) MyMethod() []Object {
	return nil
}
```

Then, after running `wails generate module`, I get the error (`wailsjs/go/models.ts`): `error TS2339: Property 'keys' does not exist on type 'typeof Object'`.

Generated code for context (focus on line 9):
```typescript
convertValues(a: any, classs: any, asMap: boolean = false): any {
	if (!a) {
		return a;
	}
	if (a.slice) {
		return (a as any[]).map(elem => this.convertValues(elem, classs));
	} else if ("object" === typeof a) {
		if (asMap) {
		     for (const key of Object.keys(a)) {
		         a[key] = new classs(a[key]);
		      }
		      return a;
		  }
		  return new classs(a);
	}
	return a;
}
```

What I think is happening here is that the JavaScript [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) is being "confused" with my generated class also named `Object`.

## Genuine questions
- Has anyone encountered similar problems?
- As a user of this project, should I adopt my code to avoid these types of conflicts? What if I am using third-party code where I have no way to edit the name of the Go structures?
- Should Wails consider giving the generated classes a prefix/suffix? (e.g. `class ObjectModel`)

**Comment by leaanthony** [comment]
Valid question! I guess you only have 2 options:
1. Avoid conflicts in naming. 
2. Generate alternate names at runtime.

I think 2 will lead to confusion as to what has been generated. 1 would be easier to deal with, however we need to improve the tooling to warn the developer that this name will be an issue. I'm assuming this is going to be true for any reserved words? We should compile a list of invalid class names and fail to build if they are invalid. It would be awesome if you could open a ticket for this! 👍

**Reply by Goncalo-Marques** [reply]
Thank you for the possible solutions 🙏 
However, I am not sure if the first option will suit my needs. Although I have ownership of the code in question, changing this structure would affect my entire project. That's why I mentioned the possibility of someone using third-party code that they can't change.

_ps: I don't want you to feel like I'm trying to impose a new feature just because something doesn't work for me... only if it makes sense for everyone._

> I'm assuming this is going to be true for any reserved words?

Thinking about this off the top of my head, JavaScript classes are being generated in Pascal Case and given that all reserved words are lowercase, I think only [built-in objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects) should be a problem. 

---

### Before I open an issue, what do you think about this:
- Adding two new flags to `wails generate module` 
   - `-prefix` (Prefix added to the generated class models)
   - `-suffix` (Suffix added to the generated class models)
- The zero values of `-prefix` and `-suffix` should work like the zero value of the string in Go
- The Pascal Case should be considered to keep consistency

#### Notes:
- We shouldn't forget to update [docs](https://wails.io/docs/reference/cli#module)
- The flags could be renamed to make them clearer

#### Example:

Generating wailsjs:
```bash
wails generate module -prefix myPrefixwithmultiplewords
```

"Pseudocode" to generate the class name:
```go
func generateClassName(prefix, suffix string) string {
   structNameGo := "Object"

   if len(prefix) != 0 {
      prefix = strings.Title(prefix)
   }
   if len(suffix) != 0 {
      suffix = strings.Title(suffix)
   }

   return prefix + structNameGo + suffix 
}
```

Result:
```javascript
class MyPrefixwithmultiplewordsObject {
}
```

### In my opinion, this solution would fix the pros and cons that we mentioned:

> What if I am using third-party code where I have no way to edit the name of the Go structures?

You can use the available flags to customize the name of the generated classes.

> I think 2 will lead to confusion as to what has been generated.

By default, the generated classes do not have any prefix or suffix.

**Reply by Goncalo-Marques** [reply]
^ Ping (@leaanthony), in case my reply was never notified 😅

**Reply by leaanthony** [reply]
Thanks for the ping. I'm ok with the prefix/suffix idea. Thoughts @stffabi ?

We need to sort out an RFC process for enhancements!

**Reply by Goncalo-Marques** [reply]
> We need to sort out an RFC process for enhancements!

Will you (Code Owners) take care of that? 
Do you need any help from me?

**Reply by leaanthony** [reply]
If you have any experience of doing RFCs from other projects, feel free to advise!

**Reply by Goncalo-Marques** [reply]
Not really 😅 
I will be watching you do this process and will take notes to help in the future, if necessary.

**Reply by stffabi** [reply]
> Thanks for the ping. I'm ok with the prefix/suffix idea. Thoughts @stffabi ?

I think that's a good idea 👍 .

**Reply by Goncalo-Marques** [reply]
Hey @leaanthony @stffabi 👋 
Any updates on the RFC process?

**Reply by leaanthony** [reply]
Not yet. Please open a ticket for adding a fixed prefix/suffix. I think the scope is narrow enough to deal with it without an RFC. Let us know how we can help any further 👍

**Reply by Goncalo-Marques** [reply]
Created issue: https://github.com/wailsapp/wails/issues/2058
